### PR TITLE
Fix camera recording issue

### PR DESCRIPTION
**Pull request by TianmiChen**
This change includes:
1. Add 16byte alignment for buffer allocation
2. Optimize buffer management with suggested
   buffer count.

Tracked-On: OAM-100215
Signed-off-by: Chen, Tianmi <tianmi.chen@intel.com>

### CODE REVIEW

**Comment by TianmiChen** [comment]
@dyang23 , @zhangyichix , @yuanjunhuang, could you please take a review ?

Thanks,

File: c2_buffers/src/mfx_c2_frame_in.cpp [hunk 1]
```diff
@@ -53,117 +53,135 @@ c2_status_t MfxC2FrameIn::init(std::shared_ptr<MfxFrameConverter> frame_converte
...
+    return res;
+}
+
+c2_status_t MfxC2FrameIn::MfxC2LoadSurfaceInSW(C2ConstGraphicBlock& c_graph_block, C2FrameData& buf_pack, c2_nsecs_t timeout)
+{
+    MFX_DEBUG_TRACE_FUNC;
+    c2_status_t res = C2_OK;
+    mfxStatus mfx_sts = MFX_ERR_NONE;
+
+	res = MapConstGraphicBlock(c_graph_block, timeout, &m_c2GraphicView);
```
Comment (zhangyichix) [inline c2_buffers/src/mfx_c2_frame_in.cpp:120]: Is there one more tab indented?

File: c2_buffers/src/mfx_c2_frame_in.cpp [hunk 2]
```diff
@@ -53,117 +53,135 @@ c2_status_t MfxC2FrameIn::init(std::shared_ptr<MfxFrameConverter> frame_converte
...
+    if(C2_OK != res) {
+        return res;
+    }
+
+	const uint8_t *pY = m_c2GraphicView->data()[C2PlanarLayout::PLANE_Y];
```
Comment (zhangyichix) [inline c2_buffers/src/mfx_c2_frame_in.cpp:125]: Is there one more tab indented?

File: c2_utils/src/mfx_defs.cpp [hunk 3]
```diff
@@ -42,57 +42,181 @@ static void InitMfxFrameHeader(
...
+    // initialize surface pool (I420, RGB4 format)
+    if (!surfpool)
+        return MFX_ERR_NULL_PTR;
+
+    uint32_t surfaceSize = MFXGetSurfaceSize(frame_info.FourCC, frame_info.Width, frame_info.Height);
+    if (!surfaceSize)
+        return MFX_ERR_MEMORY_ALLOC;
+
+    size_t framePoolBufSize = static_cast<size_t>(surfaceSize) * surfnum;
+    *buf = reinterpret_cast<uint8_t *>(calloc(framePoolBufSize, 1));
```
Comment (dyang23) [inline c2_utils/src/mfx_defs.cpp:182]: I don't think it's good idea to alloc so big buffer which cantain all surface, why don't alloc a single buffer per surface?

File: c2_utils/src/mfx_defs.cpp [hunk 4]
```diff
@@ -42,57 +42,181 @@ static void InitMfxFrameHeader(
...
+}
+
+uint32_t MFXGetFreeSurfaceIdx(mfxFrameSurface1 *SurfacesPool, uint32_t nPoolSize)
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    for (uint32_t i = 0; i < nPoolSize; i++) {
+        if (0 == SurfacesPool[i].Data.Locked)
+            return i;
+    }
```
Comment (dyang23) [inline c2_utils/src/mfx_defs.cpp:165]: this design not very flexible, this is system memory so there is no requirenment from MSDK to pre-allocate all the buffers bofore the encoder porocess start,  then we can just preallocte serverl surfaces during init status and allocate more during encode process if more surface is required?
With fixed decode surface, maybe not so much surface needed, another case more surface may needed.

File: c2_utils/src/mfx_defs.cpp (hunk 3 above)
Comment (TianmiChen) [reply c2_utils/src/mfx_defs.cpp:182]: It is just for input frames which are not 16byte aligned. If width and height are 16byte aligned, there is no need to allocate this buffer.  I have below statement to control this buffer allocation.

// 16-byte align requirement
#define MFX_C2_IS_COPY_NEEDED(_mem_type, _input_info, _mfx_info)\
    ((MFX_MEMTYPE_SYSTEM_MEMORY == _mem_type) && \
    (((_input_info).Width < (_mfx_info.Width)) || \
    ((_input_info).Height < (_mfx_info.Height))))

File: c2_utils/src/mfx_defs.cpp (hunk 4 above)
Comment (TianmiChen) [reply c2_utils/src/mfx_defs.cpp:165]: That's needed for input frames which is not 16byte aligned. Both encoder and decoder need the surface  to meet 16 byte alignment requiremnt as encoding is processed one Macroblock by one Macroblock (16x16).

**Review by dyang23 — APPROVED** [review]

**Review by zhangyichix — APPROVED** [review]

**Review by YuanjunHuang — APPROVED** [review]

**Pull request merged**